### DISCUSSION TITLE
Idea for context-aware View rendering in Spring MVC

### DISCUSSION THREAD

**Discussion by dsyer** [jstachio/jstachio#209] [Ideas]
I've been playing with a new `HandlerInterceptor` that let's you write a controller in the form:

```java
@GetMapping("/")
public DemoModel view() {
  ...
  return new DemoModel("message", thing);
}
```

where `DemoModel` is a `@JStache` model.

There is no need for `@ResponseBody` and the full `Model` is available for rendering (unlike with the `HttpMessageConverter`). The key to the implementation is that Spring automatically adds `DemoModel` to the `Model` with a well-known attribute name. The idea is that you intercept the handler response and find the `DemoModel` and turn it into a `View`, which is already supported.

IMO this is already pretty useful, but it will be even better when we get the `@context` attributes because then the view rendering can take the model into account as well.

It won't directly translate to WebFlux because there is no `HandlerInterceptor` extension point, but we already dealt with that for the JStachio module customizer thing, so it should be doable using the same approach.

I might just send a PR, but happy to clarify or discuss if you think it could be improved.

**Comment by agentgt** [comment]
You can do a PR and I understand what you are doing because I was putting something sort of like this in that is framework agnostic.

See we already have this thing called `TemplateModel` which is more or less a model and template tuple like `ModelAndView`.

If you call `jstachio.execute(templateModel)` will check if it is a `TemplateModel` and pull the real model out.

My plan is to generalize this and make a new interface called `ModelSupplier` as I have already added `ContextSupplier` https://github.com/agentgt/jstachio/blob/context_aware/api/jstachio/src/main/java/io/jstach/jstachio/context/ContextSupplier.java

Then you can make something like this:

```
public record SpringModel(Object model, ContextNode context) implements ModelSupplier, ContextSupplier {}
```

So in theory before you call JStachio but after you have retrieved the model from the model attribute you would do

```java
// write or execute
Object model = model.get(SPECIAL_ATTRIBUTE);
ContextNode context = ContextNode.of(model::get);
jstachio.write(new SpringModel(model, context), output);
```

**Comment by dsyer** [comment]
One of my goals would be for a `@Controller` not to have to know about JStachio at all - so I wouldn't want to have to import `ContextNode` or `ContextSupplier`. I assume that's still doable, but I'm not joining the dots yet on all this context-aware stuff. I will make a PR for the handler interceptor (see #210).

**Reply by agentgt** [reply]
An application writer would not have to know about ContextNode or ContextSupplier. Application users of JStachio will have to obviously know about `@context` and how it plays a little different than regular classes (as it is basically untyped).

The coupling code of `ContextNode` I mentioned would be our integration with Spring.

The only time someone would need to know about `ContextNode` is if they provided a lambda however it might not even be possible because once you open the context node stack you are stuck in there. For example let us assume I have a lambda registered on my main model called `myLambda`

```java
// method on model parent interface/class of model
@JStacheLambda 
public SomeOtherType myLambda(ContextNode contextNode){
  (SomeOtherType) contextNode.get("someOtherType");
}
```

If you need a real world example think of `SomeOtherType` as `Locale`.

```hbs
{{#@context}}
{{#myLambda}}
{{! use SomeOtherType type safely}}
{{methodOnSomeOtherType}}
{{/myLambda}}
{{/@context}}
```

Unfortunately IIRC currently JStachio will think `myLambda` is a field/method on the context when it is not (it might work but I have doubts I put that in correctly).

There are other problems with context. Like what happens if I do:

```hbs
{{@context.someThingNullOrMissing}}
```

You would think that would throw an NPE via the formatter but it does not. The reason it does not is that JStachio generates code that first checks if `someThingNullOrMissing` is `null`. This is probably bug but I can't decide.

For example we need to support:

```hbs
{{#@context.someThingNullOrMissing}}
{{.}}
{{/@context.someThingNullOrMissing}}
```

Without throwing an NPE.
